ros2param/test/test_verb_dump.py

```diff
@@ -110,52 +159,83 @@ def setUp(self):
-            assert cli.main(
-                    argv=['param', 'dump', '/foo/test_node', '--output-dir', tmpdir]) is None
-
+            with self.launch_param_dump_command(
+                arguments=[f'{TEST_NAMESPACE}/{TEST_NODE}', '--output-dir', tmpdir]
+            ) as param_dump_command:
+                assert param_dump_command.wait_for_shutdown(timeout=TEST_TIMEOUT)
+            assert param_dump_command.exit_code == launch_testing.asserts.EXIT_OK
+            assert launch_testing.tools.expect_output(
+                expected_lines=[''],
+                text=param_dump_command.output,
+                strict=True
+            )
             # Compare generated parameter file against expected
             generated_param_file = os.path.join(tmpdir, self._output_file())
             assert (open(generated_param_file, 'r').read() == EXPECTED_PARAMETER_FILE)
 
     def test_verb_dump_print(self):
-        with patch('sys.stdout', new=StringIO()) as fake_stdout:
-            assert cli.main(
-                argv=['param', 'dump', 'foo/test_node', '--print']) is None
-
-            # Compare generated stdout against expected
-            assert fake_stdout.getvalue().strip() == EXPECTED_PARAMETER_FILE[:-1]
-
+        with self.launch_param_dump_command(
+            arguments=[f'{TEST_NAMESPACE}/{TEST_NODE}', '--print']
+        ) as param_dump_command:
+            assert param_dump_command.wait_for_shutdown(timeout=TEST_TIMEOUT)
+        assert param_dump_command.exit_code == launch_testing.asserts.EXIT_OK
+        assert launch_testing.tools.expect_output(
+            expected_text=EXPECTED_PARAMETER_FILE + '\n',
+            text=param_dump_command.output,
+            strict=True
+        )
+        # If both '--output-dir' and '--print' options are provided, ensure it only prints
+        # and no file is written
         with tempfile.TemporaryDirectory() as tmpdir:
-            assert cli.main(
-                argv=['param', 'dump', 'foo/test_node', '--output-dir', tmpdir, '--print']) is None
+            with self.launch_param_dump_command(
+                arguments=[f'{TEST_NAMESPACE}/{TEST_NODE}', '--output-dir', tmpdir, '--print']
+            ) as param_dump_command:
+                assert param_dump_command.wait_for_shutdown(timeout=TEST_TIMEOUT)
+            assert param_dump_command.exit_code == launch_testing.asserts.EXIT_OK
+            assert launch_testing.tools.expect_output(
+                expected_text=EXPECTED_PARAMETER_FILE + '\n',
+                text=param_dump_command.output,
+                strict=True
+            )
 
             not_generated_param_file = os.path.join(tmpdir, self._output_file())
-
             with self.assertRaises(OSError) as context:
                 open(not_generated_param_file, 'r')
-
             # Make sure the file was not create, thus '--print' did preempt
             assert '[Errno 2] No such file or directory' in str(context.exception)
```

Comment:
@jacobperron would `os.path.exists()` and `os.path.isfile()` serve the same purpose?

Reply:
Yes, that sounds much better. This is just left over from the previous implementation, but I can change it if you like.

Reply:
Adding to the above question, is the intent to test the behavior (that no file exists) or the error message (that the exception contains '...No such file...'. I see them as somewhat distinct and I think checking the behavior is less brittle than checking the error message.

Reply:
The former. It's not so useful to check the error message, IMO, since it is not a message created by the CLI tool itself. Changed in 43701e0